### PR TITLE
ethernet: w5500: recover w5500 driver from missed interrupt

### PR DESCRIPTION
**Pull request by elcritch**
The W5500 driver will hang if a RX data interrupt is missed. This
occurs when the ethernet module is under heavy load like a flood
ping. This fix sets up the driver to timeout after some period
and manually check if the interrupt pin is active. The timeout
is configurable by setting `ETH_W5500_INT_TIMEOUT` since there is
a tradeoff between driver overhead and max incoming latency.

See similar issue and resolution at https://github.com/espressif/esp-idf/issues/6233. 

The issue has been verified and the fix tested on an nRF52840 by using flood ping. 

Signed-off-by: Jaremy J. Creechley <jaremy.creechley@panthalassa.com>

### CODE REVIEW

File: drivers/ethernet/Kconfig.w5500 [hunk 1]
```diff
@@ -35,3 +35,13 @@ config ETH_W5500_TIMEOUT
 	  Given timeout in milliseconds. Maximum amount of time
 	  that the driver will wait from the IP stack to get
 	  a memory buffer before the Ethernet frame is dropped.
+
+config ETH_W5500_INT_TIMEOUT
+	int "Interrupt gpio timeout"
+	depends on ETH_W5500
+	default 100
```
Comment (jukkar) [inline drivers/ethernet/Kconfig.w5500:42]: Should we also have some range for the value, so something like `>0 && <1000` ?

File: drivers/ethernet/Kconfig.w5500 [hunk 2]
```diff
@@ -35,3 +35,13 @@ config ETH_W5500_TIMEOUT
 	  Given timeout in milliseconds. Maximum amount of time
 	  that the driver will wait from the IP stack to get
 	  a memory buffer before the Ethernet frame is dropped.
+
+config ETH_W5500_INT_TIMEOUT
+	int "Interrupt gpio timeout"
```
Comment (jukkar) [inline drivers/ethernet/Kconfig.w5500:40]: I suggest to add the resolution of the timeout so something like `Interrupt gpio timeout in ms`

File: drivers/ethernet/Kconfig.w5500 [hunk 3]
```diff
@@ -35,3 +35,13 @@ config ETH_W5500_TIMEOUT
...
+	depends on ETH_W5500
+	default 100
+	help
+	  Maximum amount of time in milliseconds that the driver
+		will wait for an interrupt before manually checking
+		for incoming data. This adjusts how long before the driver
+		will recover from a missed interrupt.
```
Comment (jukkar) [inline drivers/ethernet/Kconfig.w5500:47]: Indentation seems to be off for the last three lines (could be also my browser that shows these wrongly)

**Review by parthitce — APPROVED** [review]
Thanks for the work. Unfortunately this puts the thread in running state even when there is no activity on ethernet. Also the way the hardware buffer is handled is impossible to check the missed data part. ACK.

**Comment by elcritch** [comment]
> Thanks for the work. Unfortunately this puts the thread in running state even when there is no activity on ethernet. 

It's unfortunate, but seems unavoidable for stable operation. Though being able to adjust the timeout should let users set it for lower overhead. On the upside, it should fix potential issues with the w5500 when waking up from sleep modes. 

> Also the way the hardware buffer is handled is impossible to check the missed data part. ACK.

We haven't noticed lost data in practice, but we're using TCP. Missing the timeout interrupt is relatively rare too, so probably not too big of a concern.

**Comment by jancoow** [comment]
I think this PR can be closed because https://github.com/zephyrproject-rtos/zephyr/pull/45765 is merged

**Pull request closed without merging**